segmentstore/storage/src/main/java/io/pravega/segmentstore/storage/chunklayer/AsyncBaseChunkStorage.java

```diff
@@ -0,0 +1,676 @@
+ *
+ * For concats, {@link ChunkStorage} supports both native and append, ChunkedSegmentStorage will invoke appropriate method depending on size of target and source chunks. (Eg. ECS)
+ *
+ * The implementations in this repository are tested using following test suites.
+ * <ul>
+ * <li>SimpleStorageTests</li>
+ * <li>ChunkedRollingStorageTests</li>
+ * <li>ChunkStorageProviderTests</li>
+ * <li>SystemJournalTests</li>
+ * </ul>
+ */
+@Slf4j
+@Beta
+public abstract class AsyncBaseChunkStorage implements ChunkStorage {
+
+    private final AtomicBoolean closed;
+
+    private final Executor executor;
+
+    /**
+     * Constructor.
+     *
+     * @param executor  An Executor for async operations.
+     */
+    public AsyncBaseChunkStorage(Executor executor) {
+        this.closed = new AtomicBoolean(false);
+        this.executor = Preconditions.checkNotNull(executor, "executor");
+    }
+
+    /**
+     * Gets a value indicating whether this Storage implementation supports truncate operation on chunks.
+     *
+     * @return True or false.
+     */
+    @Override
+    abstract public boolean supportsTruncation();
+
+    /**
+     * Gets a value indicating whether this Storage implementation supports append operation on chunks.
+     *
+     * @return True or false.
+     */
+    @Override
+    abstract public boolean supportsAppend();
+
+    /**
+     * Gets a value indicating whether this Storage implementation supports merge operation either natively or through appends.
+     *
+     * @return True or false.
+     */
+    @Override
+    abstract public boolean supportsConcat();
+
+    /**
+     * Determines whether named file/object exists in underlying storage.
+     *
+     * @param chunkName Name of the chunk to check.
+     * @return True if the object exists, false otherwise.
+     */
+    @Override
+    final public CompletableFuture<Boolean> exists(String chunkName) {
+        return executeAsync(() -> {
+            Exceptions.checkNotClosed(this.closed.get(), this);
+            // Validate parameters
+            checkChunkName(chunkName);
+
+            long traceId = LoggerHelpers.traceEnter(log, "exists", chunkName);
+            // Call concrete implementation.
+            val returnFuture =  checkExistsAsync(chunkName);
+            returnFuture.thenApplyAsync(retValue -> {
+
+                LoggerHelpers.traceLeave(log, "exists", traceId, chunkName);
+
+                return retValue;
+            });
+            return returnFuture;
+        });
+    }
+
+    /**
+     * Creates a new chunk.
+     *
+     * @param chunkName Name of the chunk to create.
+     * @return ChunkHandle A writable handle for the recently created chunk.
+     * throws ChunkStorageException Throws ChunkStorageException in case of I/O related exceptions.
+     */
+    @Override
+    final public CompletableFuture<ChunkHandle> create(String chunkName) {
+        return executeAsync(() -> {
```

Comment:
why do you need to use "executeAsync" here ?
we are only performing validations/preconditions
the real important code to be async is doCreateAsync, there is no need to add an additional layer

Reply:
Good catch. The original reason is no more valid. I removed the additional indirection.